fendermint/app/src/cmd/run.rs

```diff
@@ -52,8 +88,27 @@ async fn run(settings: Settings) -> anyhow::Result<()> {
         state_store,
         interpreter,
         resolve_pool,
+        parent_finality_provider.clone(),
     )?;
 
+    if let Some((agent_proxy, config)) = ipc_tuple {
+        let app_parent_finality_query = AppParentFinalityQuery::new(app.clone());
+        tokio::spawn(async move {
+            match launch_polling_syncer(
+                &settings.ipc.subnet_id,
+                app_parent_finality_query,
+                config,
+                parent_finality_provider,
+                agent_proxy,
+            )
```

Comment:
The initialisation above looks cool, but as we discussed in our hudl I think the polling approach is a recipe for a race condition. I am sorry, I knew I recommended in the first place. 

So my proposal is that:
* We create the provider uninitialized, as you do. 
* To initialize it is the responsibility of the `App` - specifically the Genesis interpreter or the Snapshot Loader has to call `set_committed_finality` when they learn the final height in any way. 
* On subsequent runs, the two above won't run, in which case we can do our query once - the ledger should be available. This can be done here.
* We panic if any method is called that requires init, before it's actually initialized. This should not happen. 

By polling we risk that the App receives the block to process but, alas, the polling is still asleep and the query hasn't been performed.

Reply:
I think adding to `Genesis Interpreter` might introduce extra code burden as we have to introduce `ipc-agent` related code to get the subnet genesis epoch. Also the ChainInterpreter init chain does not have the AppState initialized. Maybe we add to `App::init_chain`, but somehow, we have to deal with the extra imported crate the `App` needs to hold. 

I feel the ideal case would be starting the `Polling` in the background then we run the `App`. But because we need to query the `AppState` that we have to init the `App` first. We can add a barrier  in `App` to signal start the service only when `Polling` has started. That way we dont pollute `App` with too many dependencies, but also starting the `App` only when `Polling` is running.

Btw, I'm wondering if the race is indeed an issue. Because the polling is not ready, that means we dont have anything, maybe dont propose and we dont vote for any proposals. Does this impact `tendermint` a lot?

Reply:
Thanks, I can't find it now but I commented somewhere that I agree that because you are never returning errors due to "data not ready" any more, polling is fine, we just skip a round of proposals, no biggie, no race :+1: